src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs

```diff
@@ -244,6 +255,20 @@ internal ValueTask StartAsync(CancellationToken cancellationToken = default)
                 int status = MsQuicApi.Api.StreamStart(
                     _handle,
                     QUIC_STREAM_START_FLAGS.SHUTDOWN_ON_FAIL | QUIC_STREAM_START_FLAGS.INDICATE_PEER_ACCEPT);
+
+                if (status == QUIC_STATUS_ABORTED)
+                {
+                    // Connection has been closed by the peer (either at transport or application level),
```

Comment:
`QUIC_STREAM_START_FLAGS.SHUTDOWN_ON_FAIL` means that we should get shutdown. If we don't, perhaps this is a bug in MsQuic. At least I'd interested in @nibanks opinion on this.

Reply:
From glance at source code, the SHUTDOWN_ON_FAIL applies only if the `MsQuicStreamStart` finishes with the `QUIC_STATUS_PENDING`, and the operation fails (e.g. due to connection getting closed meanwhile) when executed from the worker thread.

Reply:
Radek is right that the first requirement is that StreamStart succeeds (pending). If any other result is returned, then you won't be getting any further notification. But if we are able to queue the start operation you must always get a notification when you use the `QUIC_STREAM_START_FLAG_SHUTDOWN_ON_FAIL` flag. There are two places it may occur in the QUIC stack:

1. In the normal stream start path (`QuicStreamStart`) if there is any failure (i.e. OOM or `FAIL_ON_BLOCKED` is used).
2. In the connection operation queue clean up path for race conditions where the connection is cleaning up (`QuicOperationQueueClear`).

If you have an example that doesn't follow this, please grab some logs and share. Thanks!